File: dart-language/charge-system-demo/client-ng-dart/lib/src/user_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart';
import 'uri_service.dart';

class UserService {
  static final loginUrl = 'api/login';
  static final _headers = { 'content-type': 'applicationo/json'};
  static String currentUser = '';
  static bool auth = false;
  final Client _http;
  final UriService uriService;

  UserService(this._http, this.uriService);

  Exception _handleError(dynamic e) {
    print(e);
    return Exception('Server error: cause: $e');
  }
  Future<bool> login(String username, String password)  async {
    try {
      Uri url = uriService.formatUri(loginUrl);
      final res = await _http.post(url, headers: _headers, body: json.encode({'username': username, 'password': password}));
      currentUser = username;
      auth = true;
      return true;
    } catch (e) {
      throw _handleError(e);
    }
  }
  bool logout() {
    currentUser = '';
    auth = false;
    return true;
  }
  bool isAuthed() {
    return auth;
  }
  String getCurrentUser() {
    return currentUser;
  }
}

File: dart-language/charge-system-demo/client-ng-ts/projects/at-antd/src/app/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError, map, tap } from 'rxjs/operators';
import { NzMessageService } from 'ng-zorro-antd';

const httpOptions = {
  headers: new HttpHeaders({ 'content-type': 'application/json' })
};

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private loginUrl = 'api/login';
  private currentUser = '';
  private auth = false;

  constructor(private http: HttpClient, private message: NzMessageService) { }

  private handleError<T>(operation = 'operation', result?: T) {
    return (error: any): Observable<T> => {
      this.message.create('error', `${operation} failed: ${error.error.message}`);
      return of(result);
    }
  }

  login(username, password): Observable<any> {
    const body = JSON.stringify({ username, password });
    return this.http.post(this.loginUrl, body, httpOptions).pipe(
      tap(() => { this.setUsername(username); this.setAuth(true) }),
      map((res) => res['data']),
      catchError(this.handleError<any>('login'))
    );
  }
  logout() {
    this.setUsername();
    this.setAuth();
  }

  setUsername(name: string = ''): void {
    this.currentUser = name;
  }
  setAuth(auth: boolean = false) {
    this.auth = auth;
  }
  isAuthed(): boolean {
    return this.auth;
  }
  getCurrentUser(): string {
    return this.currentUser;
  }
}
